README
This package requires the Fortran LAPACK library version 3.0 available
in source form and in compiled form for some systems at
		    http://www.netlib.org/lapack/
The configure script for this package attempts to detect the lapack
library and the Atlas or BLAS libraries.  Please tell us if this does
not work for you.

Douglas Bates <bates@stat.wisc.edu>
Saikat DebRoy <saikat@stat.wisc.edu>



ChangeLog
2015-04-01  Martin Maechler  <maechler@stat.math.ethz.ch>

	* inst/NEWS.Rd: Moving ChangeLog to new (markup) NEWS file

2015-03-16  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/Auxiliaries.R (anyDuplicatedT): renamed from is_duplicatedT(),
	and exported.  uniqTsparse(): exported too.

2014-06-15  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.1-4, released to CRAN on 2014-06-14
	* src/dsyMatrix.c (dsyMatrix_matrix_mm): fix crossprod(<vec>, <dsyMatrix>) bug

2014-04-26  Martin Maechler  <maechler@stat.math.ethz.ch>

	* new rsparsematrix()

2014-03-30  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.1-3, released to CRAN on 2014-03-30

2014-03-12  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/dgTMatrix.R (image): fix bug in default ylim computation.

2014-01-28  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/products.R: matrix products overhauled; should work with sparseVectors;
	speedup of crossprod(v, <sparseM>), thanks to nudge by Niels Richard Hansen.
	* man/matrix-products.Rd: all matrix products documented in one file.
	* tests/matprod.R: more extensive testing

2014-01-20  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.1-2, released to CRAN on 2014-01-28

	* NAMESPACE: export fast power-user coercion utilities
	.dsy2mat(), .dxC2mat(), .T2Cmat(), ..2dge().

2013-12-23  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/dgTMatrix.R (image): (xlim, ylim) get a slightly changed
	default, plus  ylim := sort(ylim, "decreasing").
	This is strictly not back-compatible but should never harm.

2013-09-26  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/spModels.R (fac2sparse, fac2Sparse): newly exported; plus
	'giveCsparse' option.

2013-09-16  Martin Maechler  <maechler@stat.math.ethz.ch>

	* src/scripts/0get-SuiteSparse.sh: new download script, for
	* src/CHOLMOD/*, src/AMD/*, ...: getting SuiteSparse version 4.2.1

	* R/zzz.R (.SuiteSparse_version): new function

2013-09-13  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/dsCMatrix.R (solve.dsC.*): finally fix the long-lasting
	undetected solve(<dsC>) bug (only in case Cholmod fails) [r2908].

	* DESCRIPTION (Version): 1.0-15, CRAN-*non*-released 2013-09-26

2013-09-12  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-14, CRAN-released 2013-09-12
	* R/dgCMatrix.R: "stop gap fix" for R 3.0.2: partly revert
	solve(<sparse>,*) changes in 1.0-13.

2013-08-27 Fabian Scheipl  <fabian.scheipl@stat.uni-muenchen.de>

	* man/indMatrix-class.Rd: new "indMatrix" class, a natural
	superclass of "pMatrix".  Many methods moved from "pMatrix" to
	"indMatrix".

2013-05-09  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-13, CRAN-released 2013-09-10

	* R/KhatriRao.R: Efficient KhatriRao() by Michael Cysouw

2013-03-26  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-12, CRAN-released 2013-03-26

2012-11-10  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-11, CRAN-released 2013-02-02

	* R/SparseM-conv.R: as(<csr>, "dgCMatrix") now works again
	* tests/other-pkgs.R: test that.

	* src/Mutils.c: do *not* use '#if R_VERSION < ..' so this also
	*runs* in older R when installed in R >= 2.15.2.

2012-10-15  Martin Maechler  <maechler@stat.math.ethz.ch>

	* src/Mutils.c (Mmatrix): new, to be used by .External()
	in order to replace .Internal(matrix(....))

2012-10-05  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/diagMatrix.R (.sparseDiagonal): new 'unitri' argument;
	more flexibility; new: solve(<dsCMatrix>, <missing>)

2012-09-10  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-10, CRAN-released: 2012-10-16, r2845

2012-09-01  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/sparseVector.R (sparseVector): new constructor

	* inst/test-tools-Matrix.R (rspMat): smarter; also useful for large
	dimensions.

2012-07-23  Martin Maechler  <maechler@stat.math.ethz.ch>

	* tests/group-methods.R: now do much more testing, notably <Ops> of
	pairs of matrices ... however only when 'doExtras' is true, which
	it is not by default, e.g., on CRAN.

2012-07-21  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/Ops.R, R/diagMatrix.R: many fixes, notably for rare operations
	that were not triggered before.

	* R/Auxiliaries.R (allTrueMat): new utility.

2012-07-20  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/dsparseMatrix.R, R/sparseVector.R, ...: Newly defined
	is.finite() and is.infinite() methods for all our *Matrix and *sparseVector.

2012-07-14  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/d??Matrix.R (diag<-): many "diag<-" methods, which preserve
	symmetricity, triangularity (even uni-triangularity sometimes);
	Partly by also making   A[cbind(i,i)]  preserve such properties.

	* src/Mutils.c (SET_packed_setDiag) {and similar}:
	*_setDiag() C functions implementing "diag<-" R methods.

2012-06-30  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/Ops.R (Ops.x.x): now, dense symmetric and triangular matrices
	are preserved for many arithmetic and logic binary operators.

	* src/ldense.c (lsyMatrix_as_lspMatrix, ..): more coercions keep
	the dimnames().

	* R/symmetricMatrix.R (pack, unpack): new pack() {"inverse" of
	unpack(), including pack(<matrix>)}; new unpack() methods.

2012-06-20  Douglas Bates <bates@stat.wisc.edu>

	* src/scripts/DEPS.mkf,	...
	* src/*.c: Update to version 4.0.0 of SuiteSparse

2012-06-19  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-8, CRAN-released: 2012-06-20, r2789

	* R/CHMfactor.R (update): *DO* allow non-symmetric parent.
	* man/CHMfactor-class.Rd: be more clear about the two ways.
	* tests/factorizing.R: more update(<Chol>) testing

2012-06-12  Martin Maechler  <maechler@stat.math.ethz.ch>

	* tests/matprod.R (chkDnProd): new testing function

2012-06-08  Martin Maechler  <maechler@stat.math.ethz.ch>


R/HBMM.R
## Utilities for the Harwell-Boeing and MatrixMarket formats

readone <- function(ln, iwd, nper, conv)
{
    ln <- gsub("D", "E", ln)
    inds <- seq(0, by = iwd, length = nper + 1)
    (conv)(substring(ln, 1 + inds[-length(inds)], inds[-1]))
}

readmany <- function(conn, nlines, nvals, fmt, conv)
{
    if (!grep("[[:digit:]]+[DEFGI][[:digit:]]+", fmt))
	stop("Not a valid format")
    Iind <- regexpr('[DEFGI]', fmt)
    nper <- as.integer(substr(fmt, regexpr('[[:digit:]]+[DEFGI]', fmt), Iind - 1))
    iwd <- as.integer(substr(fmt, Iind + 1, regexpr('[\\.\\)]', fmt) - 1))
    rem <- nvals %% nper
    full <- nvals %/% nper
    ans <- vector("list", nvals %/% nper)
    for (i in seq_len(full))
	ans[[i]] <- readone(readLines(conn, 1, ok = FALSE),
			    iwd, nper, conv)
    if (!rem) return(unlist(ans))
    c(unlist(ans),
      readone(readLines(conn, 1, ok = FALSE), iwd, rem, conv))
}

readHB <- function(file)
{
    if (is.character(file))
	file <- if (file == "") stdin() else file(file)
    if (!inherits(file, "connection"))
        stop("'file' must be a character string or connection")
    if (!isOpen(file)) {
        open(file)
        on.exit(close(file))
    }
    hdr <- readLines(file, 4, ok = FALSE)
    Title <- sub('[[:space:]]+$', '', substr(hdr[1], 1, 72))
    Key <- sub('[[:space:]]+$', '', substr(hdr[1], 73, 80))
    totln <- as.integer(substr(hdr[2], 1, 14))
    ptrln <- as.integer(substr(hdr[2], 15, 28))
    indln <- as.integer(substr(hdr[2], 29, 42))
    valln <- as.integer(substr(hdr[2], 43, 56))
    rhsln <- as.integer(substr(hdr[2], 57, 70))
    if (!(t1 <- substr(hdr[3], 1, 1)) %in% c('C', 'R', 'P'))
	stop("Invalid storage type: ", t1)
    if (t1 != 'R') stop("Only numeric sparse matrices allowed")
    ## _FIXME: Patterns should also be allowed
    if (!(t2 <- substr(hdr[3], 2, 2)) %in% c('H', 'R', 'S', 'U', 'Z'))
	stop("Invalid storage format: ", t2)
    if (!(t3 <- substr(hdr[3], 3, 3)) %in% c('A', 'E'))
	stop("Invalid assembled indicator: ", t3)
    nr <- as.integer(substr(hdr[3], 15, 28))
    nc <- as.integer(substr(hdr[3], 29, 42))
    nz <- as.integer(substr(hdr[3], 43, 56))
    nel <- as.integer(substr(hdr[3], 57, 70))
    ptrfmt <- toupper(sub('[[:space:]]+$', '', substr(hdr[4], 1, 16)))
    indfmt <- toupper(sub('[[:space:]]+$', '', substr(hdr[4], 17, 32)))
    valfmt <- toupper(sub('[[:space:]]+$', '', substr(hdr[4], 33, 52)))
    rhsfmt <- toupper(sub('[[:space:]]+$', '', substr(hdr[4], 53, 72)))
    if (!is.na(rhsln) && rhsln > 0) {
        h5 <- readLines(file, 1, ok = FALSE)
    }
    ptr <- readmany(file, ptrln, nc + 1, ptrfmt, as.integer)
    ind <- readmany(file, indln, nz, indfmt, as.integer)
    vals <- readmany(file, valln, nz, valfmt, as.numeric)
    if (t2 == 'S')
        new("dsCMatrix", uplo = "L", p = ptr - 1L,
            i = ind - 1L, x = vals, Dim = c(nr, nc))
    else
        new("dgCMatrix", p = ptr - 1L,
            i = ind - 1L, x = vals, Dim = c(nr, nc))

}

readMM <- function(file)
{
    if (is.character(file))
	file <- if(file == "") stdin() else file(file)
    if (!inherits(file, "connection"))
	stop("'file' must be a character string or connection")
    if (!isOpen(file)) {
	open(file)
	on.exit(close(file))
    }
    scan1 <- function(what, ...)
	scan(file, nmax = 1, what = what, quiet = TRUE, ...)

    if ((hdr <- scan1(character())) != "%%MatrixMarket")
	stop("file is not a MatrixMarket file")
    if (!(typ <- tolower(scan1(character()))) %in% "matrix")
	stop(gettextf("type '%s' not recognized", typ), domain = NA)
    if (!(repr <- tolower(scan1(character()))) %in% c("coordinate", "array"))
	stop(gettextf("representation '%s' not recognized", repr), domain = NA)
    elt <- tolower(scan1(character()))
    if (!elt %in% c("real", "complex", "integer", "pattern"))
	stop(gettextf("element type '%s' not recognized", elt), domain = NA)

    sym <- tolower(scan1(character()))
    if (!sym %in% c("general", "symmetric", "skew-symmetric", "hermitian"))
	stop(gettextf("symmetry form '%s' not recognized", sym), domain = NA)
    nr <- scan1(integer(), comment.char = "%")
    nc <- scan1(integer())
    nz <- scan1(integer())
    checkIJ <- function(els) {
	if(els$i < 1 || els$i > nr)
	    stop("readMM(): row	 values 'i' are not in 1:nr", call.=FALSE)
	if(els$j < 1 || els$j > nc)
	    stop("readMM(): column values 'j' are not in 1:nc", call.=FALSE)
    }
    if (repr == "coordinate") {
	switch(elt,
	       "real" = ,
	       "integer" = {
		   ## TODO: the "integer" element type should be returned as
		   ##       an object of an "iMatrix" subclass--once there are
		   els <- scan(file, nmax = nz, quiet = TRUE,
			       what= list(i= integer(), j= integer(), x= numeric()))
                   checkIJ(els)
		   switch(sym,
			  "general" = {
			      new("dgTMatrix", Dim = c(nr, nc), i = els$i - 1L,
				  j = els$j - 1L, x = els$x)
			  },
			  "symmetric" = {
			      new("dsTMatrix", uplo = "L", Dim = c(nr, nc),
				  i = els$i - 1L, j = els$j - 1L, x = els$x)
			  },
			  "skew-symmetric" = {
			      stop("symmetry form 'skew-symmetric' not yet implemented for reading")
			      ## FIXME: use dgT... but must expand the (i,j,x) slots!
			      new("dgTMatrix", uplo = "L", Dim = c(nr, nc),
				  i = els$i - 1L, j = els$j - 1L, x = els$x)

			  },
			  "hermitian" = {
			      stop("symmetry form 'hermitian' not yet implemented for reading")
			  },
			  ## otherwise (not possible; just defensive programming):
			  stop(gettextf("symmetry form '%s' is not yet implemented",
					sym), domain = NA)
			  )
	       },
	       "pattern" = {
		   els <- scan(file, nmax = nz, quiet = TRUE,
			       what = list(i = integer(), j = integer()))
		   checkIJ(els)
		   switch(sym,
			  "general" = {
			      new("ngTMatrix", Dim = c(nr, nc),
				  i = els$i - 1L, j = els$j - 1L)
			  },
			  "symmetric" = {
			      new("nsTMatrix", uplo = "L", Dim = c(nr, nc),
				  i = els$i - 1L, j = els$j - 1L)
			  },
			  "skew-symmetric" = {
			      stop("symmetry form 'skew-symmetric' not yet implemented for reading")
			      ## FIXME: use dgT... but must expand the (i,j,x) slots!
			      new("ngTMatrix", uplo = "L", Dim = c(nr, nc),
				  i = els$i - 1L, j = els$j - 1L)

			  },
			  "hermitian" = {
			      stop("symmetry form 'hermitian' not yet implemented for reading")
			  },
			  ## otherwise (not possible; just defensive programming):
			  stop(gettextf("symmetry form '%s' is not yet implemented",
					sym), domain = NA)
			  )
	       },
	       "complex" = {
		   stop("element type 'complex' not yet implemented")
	       },
	       ## otherwise (not possible currently):
	       stop(gettextf("'%s()' is not yet implemented for element type '%s'",
			     "readMM", elt), domain = NA))
    }
    else
	stop(gettextf("'%s()' is not yet implemented for  representation '%s'",
		      "readMM", repr), domain = NA)
}
